www/squid/files/extra-patch-host-header-forgery

```diff
@@ -1,12 +1,28 @@
---- src/client_side_request.cc.orig	2019-02-19 02:46:22 UTC
+--- src/client_side_request.cc.orig	2022-09-05 06:06:48.000000000 +0200
 +++ src/client_side_request.cc
-@@ -538,6 +538,9 @@ ClientRequestContext::hostHeaderIpVerify
-             }
-             debugs(85, 3, HERE << "validate IP " << clientConn->local << " non-match from Host: IP " << ia->in_addrs[i]);
-         }
-+        http->request->flags.hostVerified = true;
-+        http->doCallouts();
-+        return;
+@@ -524,17 +524,22 @@
+ ClientRequestContext::hostHeaderIpVerify(const ipcache_addrs* ia, const Dns::LookupDetails &dns)
+ {
+     Comm::ConnectionPointer clientConn = http->getConn()->clientConnection;
+ 
+     // note the DNS details for the transaction stats.
+     http->request->recordLookup(dns);
+ 
+-    // Is the NAT destination IP in DNS?
+-    if (ia && ia->have(clientConn->local)) {
+-        debugs(85, 3, "validate IP " << clientConn->local << " possible from Host:");
++    if (ia) {
```

Comment:
reviewing a diff in a diff is a bit difficult, but it looks like a change in behaviour? or did squid also change the line where this was supposed to apply making the fuzzing match even more odd.

Reply:
Yes, sorry of the diff in diff. Here's something more readable (or a [direct link](https://github.com/marinbernard-pep06/opnsense-ports/blob/c6ab2f03a91c00798193afd1c6f2c31bd2eed5d9/www/squid/files/extra-patch-host-header-forgery)):

```
--- src/client_side_request.cc.orig	2022-09-05 06:06:48.000000000 +0200
+++ src/client_side_request.cc
@@ -524,17 +524,22 @@
 ClientRequestContext::hostHeaderIpVerify(const ipcache_addrs* ia, const Dns::LookupDetails &dns)
 {
     Comm::ConnectionPointer clientConn = http->getConn()->clientConnection;
 
     // note the DNS details for the transaction stats.
     http->request->recordLookup(dns);
 
-    // Is the NAT destination IP in DNS?
-    if (ia && ia->have(clientConn->local)) {
-        debugs(85, 3, "validate IP " << clientConn->local << " possible from Host:");
+    if (ia) {
+        // Is the NAT destination IP in DNS?
+        // Do not enforce IP validation on disabled host-header forgery check
+        if (ia->have(clientConn->local)) {
+            debugs(85, 3, "validate IP " << clientConn->local << " possible from Host:");
+        } else {
+            debugs(85, 3, "validate IP " << clientConn->local << " skipped from Host:");
+        }
         http->request->flags.hostVerified = true;
         http->doCallouts();
         return;
     }
     debugs(85, 3, HERE << "FAIL: validate IP " << clientConn->local << " possible from Host:");
     hostHeaderVerifyFailed("local IP", "any domain IP");
 }
```

Yes, the function was rewritten and simplified. Basically, the logic trying to match the IP address supplied by the client with an IP address resolved by the proxy was moved to another function (invoked with `ia->have()`). So in the end the [v5][v5] version is simpler than in [v4][v4].

I could have made a simpler patch by simply replacing the last 2 lines of the function with:

```
    http->request->flags.hostVerified = true;
    http->doCallouts();
```

But I wanted to log a message when the host header IP validation was skipped due to he patch, so I split the conditional in two. I also added 7 lines of context (instead of 3 by default) to avoid patching again the wrong region.

[v4]: https://github.com/squid-cache/squid/blob/v4/src/client_side_request.cc#L523
[v5]: https://github.com/squid-cache/squid/blob/v5/src/client_side_request.cc#L523